src/renderer/src/pass/forward.rs

```diff
@@ -69,72 +170,213 @@ impl<R> Pass<R> for Clear
+    fn apply<C>(&self, arg: &pass::DrawFlat, target: &ColorBuffer<R>, scenes: &::Frame<R>, encoder: &mut gfx::Encoder<R, C>)
         where C: gfx::CommandBuffer<R>
     {
         let scene = &scenes.scenes[&arg.scene];
         let camera = &scenes.cameras[&arg.camera];
 
         // every entity gets drawn
         for e in &scene.fragments {
+            encoder.update_constant_buffer(
+                &self.vertex,
+                &VertexArgs{
+                    proj: camera.projection,
+                    view: camera.view,
+                    model: e.transform,
+                }
+            );
+
+            encoder.update_constant_buffer(
+                &self.fragment,
+                &FragmentArgs{
+                    light_count: 0
+                }
+            );
+
+            let ka = e.ka.to_view(&self.ka, encoder);
+            let kd = e.kd.to_view(&self.kd, encoder);
+
             encoder.draw(
                 &e.slice,
-                &self.0,
+                &self.pso,
                 &flat::Data{
                     vbuf: e.buffer.clone(),
-                    ka: e.ka,
-                    model: e.transform,
-                    view: camera.view,
-                    proj: camera.projection,
+                    vertex_args: self.vertex.clone(),
+                    fragment_args: self.fragment.clone(),
                     out_ka: target.color.clone(),
-                    out_depth: target.output_depth.clone()
+                    out_depth: target.output_depth.clone(),
+                    ka: (ka, self.sampler.clone()),
+                    kd: (kd, self.sampler.clone()),
                 }
             );
         }
     }
 }
 
-gfx_pipeline!( wireframe {
-    vbuf: gfx::VertexBuffer<VertexPosNormal> = (),
-    ka: gfx::Global<[f32; 4]> = "u_Ka",
-    model: gfx::Global<[[f32; 4]; 4]> = "u_Model",
-    view: gfx::Global<[[f32; 4]; 4]> = "u_View",
-    proj: gfx::Global<[[f32; 4]; 4]> = "u_Proj",
-    out_ka: gfx::RenderTarget<gfx::format::Rgba8> = "o_Ka",
-});
+pub struct DrawShaded<R: gfx::Resources>{
+    vertex: gfx::handle::Buffer<R, VertexArgs>,
+    fragment: gfx::handle::Buffer<R, FragmentArgs>,
+    lights: gfx::handle::Buffer<R, PointLight>,
+    pso: gfx::pso::PipelineState<R, shaded::Meta>,
+    sampler: gfx::handle::Sampler<R>,
+    ka: ::ConstantColorTexture<R>,
+    kd: ::ConstantColorTexture<R>,
+}
+
+impl<R: gfx::Resources> DrawShaded<R> {
+    pub fn new<F>(factory: &mut F) -> DrawShaded<R>
+        where R: gfx::Resources,
+              F: gfx::Factory<R>
+    {
+        let lights = factory.create_constant_buffer(512);
+        let vertex = factory.create_constant_buffer(1);
+        let fragment = factory.create_constant_buffer(1);
+        let pso = factory.create_pipeline_simple(
+            VERTEX_SRC,
+            FRAGMENT_SRC,
+            shaded::new()
+        ).unwrap();
+
+        let sampler = factory.create_sampler(
```

Comment:
seeing this repeated a lot in our code (not just in Amethyst), we should make this more ergonomic